File: app/src/main/java/org/alkaline/taskbrain/data/ContentSimilarity.kt
package org.alkaline.taskbrain.data

/**
 * Content similarity utilities for matching edited lines to their originals.
 *
 * Used by line matching algorithms (EditorState.updateFromText, CurrentNoteViewModel)
 * to assign note IDs to the line fragment with the highest proportion of original
 * content after a line split.
 */

/**
 * Returns the proportion of [oldContent] that appears in [newContent],
 * measured by longest common subsequence length / old content length.
 * Returns 0.0 early if the strings share no characters.
 */
fun contentOverlapProportion(oldContent: String, newContent: String): Double {
    if (oldContent.isEmpty()) return 0.0
    if (!sharesAnyCharacter(oldContent, newContent)) return 0.0
    return lcsLength(oldContent, newContent).toDouble() / oldContent.length.toDouble()
}

private fun sharesAnyCharacter(a: String, b: String): Boolean {
    val charSet = a.toSet()
    return b.any { it in charSet }
}

/**
 * Computes the length of the longest common subsequence between two strings.
 * Space-optimized to O(min(m, n)).
 */
fun lcsLength(a: String, b: String): Int {
    if (a.isEmpty() || b.isEmpty()) return 0
    val m = a.length
    val n = b.length
    var prev = IntArray(n + 1)
    var curr = IntArray(n + 1)
    for (i in 1..m) {
        for (j in 1..n) {
            curr[j] = if (a[i - 1] == b[j - 1]) prev[j - 1] + 1
            else maxOf(prev[j], curr[j - 1])
        }
        val temp = prev
        prev = curr
        curr = temp
        curr.fill(0)
    }
    return prev[n]
}

/**
 * Matches unmatched new lines to unconsumed old lines by content similarity.
 *
 * For each (old, new) pair, computes the proportion of old content present in the
 * new content (via LCS). Greedily assigns the highest-proportion match first, ensuring
 * that when a line is split, the fragment with more original content keeps the ID.
 *
 * @param oldContents text of old lines
 * @param newContents text of new lines
 * @param oldConsumed which old lines are already matched; updated in place
 * @param newMatched which new lines are already matched (non-null entries); updated in place
 * @param getOldContent extracts the text to compare from old line index
 * @param getNewContent extracts the text to compare from new line index
 */
fun performSimilarityMatching(
    unmatchedNewIndices: Set<Int>,
    unconsumedOldIndices: List<Int>,
    getOldContent: (Int) -> String,
    getNewContent: (Int) -> String,
    onMatch: (oldIdx: Int, newIdx: Int) -> Unit
) {
    data class Match(val oldIdx: Int, val newIdx: Int, val proportion: Double)

    val candidates = mutableListOf<Match>()
    for (oldIdx in unconsumedOldIndices) {
        val oldContent = getOldContent(oldIdx)
        for (newIdx in unmatchedNewIndices) {
            val proportion = contentOverlapProportion(oldContent, getNewContent(newIdx))
            if (proportion > 0.0) {
                candidates.add(Match(oldIdx, newIdx, proportion))
            }
        }
    }
    candidates.sortByDescending { it.proportion }
    val matchedOld = mutableSetOf<Int>()
    val matchedNew = mutableSetOf<Int>()
    for (match in candidates) {
        if (match.oldIdx in matchedOld || match.newIdx in matchedNew) continue
        onMatch(match.oldIdx, match.newIdx)
        matchedOld.add(match.oldIdx)
        matchedNew.add(match.newIdx)
    }
}

/**
 * Determines which half of a split line should keep the noteIds.
 *
 * When [noteIdContentLengths] is available (from a prior merge), each noteId is assigned
 * to the half that contains more of its original content. This correctly distributes
 * noteIds back to their original lines after a merge–split round-trip.
 *
 * When content lengths aren't available (single noteId or loaded from server),
 * all noteIds go to the longer half.
 *
 * @param noteIds the line's noteIds (in text order when from a merge)
 * @param beforeContentLen length of content before the split point (excluding prefix)
 * @param afterContentLen length of content after the split point
 * @param beforeHasContent whether the before half has any content
 * @param afterHasContent whether the after half has any content
 * @param noteIdContentLengths per-noteId content lengths in text order (from merge metadata)
 * @return pair of (beforeNoteIds, afterNoteIds)
 */
fun splitNoteIds(
    noteIds: List<String>,
    beforeContentLen: Int,
    afterContentLen: Int,
    beforeHasContent: Boolean,
    afterHasContent: Boolean,
    noteIdContentLengths: List<Int> = emptyList()
): Pair<List<String>, List<String>> {
    val (before, after) = when {
        // One or no content side — all noteIds go to the side with content
        !beforeHasContent && afterHasContent -> emptyList<String>() to noteIds
        beforeHasContent && !afterHasContent -> noteIds to emptyList()
        // Multiple noteIds with content-length metadata — distribute by overlap
        noteIds.size > 1 && noteIdContentLengths.size == noteIds.size ->
            distributeNoteIdsByOverlap(noteIds, beforeContentLen, noteIdContentLengths)
        // Fallback: all noteIds go to the longer half
        beforeContentLen >= afterContentLen -> noteIds to emptyList()
        else -> emptyList<String>() to noteIds
    }
    // Any side that ended up with content but no id gets a SPLIT sentinel, so
    // save-time attribution ("where did this fresh doc come from?") is consistent.
    return stampSplitSentinelIfNeeded(before, beforeHasContent) to
        stampSplitSentinelIfNeeded(after, afterHasContent)
}

private fun stampSplitSentinelIfNeeded(ids: List<String>, hasContent: Boolean): List<String> =
    if (ids.isEmpty() && hasContent) listOf(NoteIdSentinel.new(NoteIdSentinel.Origin.SPLIT)) else ids

/**
 * Distributes noteIds to before/after halves based on how much of each noteId's
 * original content falls on each side of the split point.
 */
private fun distributeNoteIdsByOverlap(
    noteIds: List<String>,
    splitPos: Int,
    contentLengths: List<Int>
): Pair<List<String>, List<String>> {
    val beforeIds = mutableListOf<String>()
    val afterIds = mutableListOf<String>()
    var offset = 0

    for (i in noteIds.indices) {
        val len = contentLengths[i]
        val end = offset + len
        val overlapBefore = (minOf(end, splitPos) - offset).coerceAtLeast(0)
        val overlapAfter = (end - maxOf(offset, splitPos)).coerceAtLeast(0)

        if (overlapBefore >= overlapAfter) {
            beforeIds.add(noteIds[i])
        } else {
            afterIds.add(noteIds[i])
        }
        offset = end
    }

    return beforeIds to afterIds
}

File: web/src/editor/ContentSimilarity.ts
/**
 * Content similarity utilities for matching edited lines to their originals.
 *
 * Used by line matching algorithms (NoteRepository.matchLinesToIds, EditorState.updateFromText)
 * to assign note IDs to the line fragment with the highest proportion of original content
 * after a line split.
 */

import { newSentinelNoteId } from '@/data/NoteIdSentinel'

/**
 * Returns the proportion of oldContent that appears in newContent,
 * measured by longest common subsequence length / old content length.
 * Returns 0 early if the strings share no characters.
 */
export function contentOverlapProportion(oldContent: string, newContent: string): number {
  if (oldContent.length === 0) return 0
  if (!sharesAnyCharacter(oldContent, newContent)) return 0
  return lcsLength(oldContent, newContent) / oldContent.length
}

function sharesAnyCharacter(a: string, b: string): boolean {
  const charSet = new Set(a)
  for (const ch of b) {
    if (charSet.has(ch)) return true
  }
  return false
}

/**
 * Computes the length of the longest common subsequence between two strings.
 * Space-optimized to O(min(m, n)).
 */
export function lcsLength(a: string, b: string): number {
  if (a.length === 0 || b.length === 0) return 0
  const m = a.length
  const n = b.length
  let prev = new Array<number>(n + 1).fill(0)
  let curr = new Array<number>(n + 1).fill(0)
  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      curr[j] =
        a[i - 1] === b[j - 1] ? prev[j - 1]! + 1 : Math.max(prev[j]!, curr[j - 1]!)
    }
    const temp = prev
    prev = curr
    curr = temp
    curr.fill(0)
  }
  return prev[n]!
}

/**
 * Matches unmatched new lines to unconsumed old lines by content similarity.
 *
 * For each (old, new) pair, computes the proportion of old content present in the
 * new content (via LCS). Greedily assigns the highest-proportion match first, ensuring
 * that when a line is split, the fragment with more original content keeps the ID.
 */
export function performSimilarityMatching(
  unmatchedNewIndices: Set<number>,
  unconsumedOldIndices: number[],
  getOldContent: (idx: number) => string,
  getNewContent: (idx: number) => string,
  onMatch: (oldIdx: number, newIdx: number) => void,
): void {
  interface Match {
    oldIdx: number
    newIdx: number
    proportion: number
  }

  const candidates: Match[] = []
  for (const oldIdx of unconsumedOldIndices) {
    const oldContent = getOldContent(oldIdx)
    for (const newIdx of unmatchedNewIndices) {
      const proportion = contentOverlapProportion(oldContent, getNewContent(newIdx))
      if (proportion > 0) {
        candidates.push({ oldIdx, newIdx, proportion })
      }
    }
  }
  candidates.sort((a, b) => b.proportion - a.proportion)
  const matchedOld = new Set<number>()
  const matchedNew = new Set<number>()
  for (const match of candidates) {
    if (matchedOld.has(match.oldIdx) || matchedNew.has(match.newIdx)) continue
    onMatch(match.oldIdx, match.newIdx)
    matchedOld.add(match.oldIdx)
    matchedNew.add(match.newIdx)
  }
}

/**
 * Determines which half of a split line should keep the noteIds.
 *
 * When noteIdContentLengths is available (from a prior merge), each noteId is assigned
 * to the half that contains more of its original content. This correctly distributes
 * noteIds back to their original lines after a merge–split round-trip.
 *
 * @returns [beforeNoteIds, afterNoteIds]
 */
export function splitNoteIds(
  noteIds: string[],
  beforeContentLen: number,
  afterContentLen: number,
  beforeHasContent: boolean,
  afterHasContent: boolean,
  noteIdContentLengths: number[] = [],
): [string[], string[]] {
  let before: string[]
  let after: string[]
  if (!beforeHasContent && afterHasContent) { before = []; after = noteIds }
  else if (beforeHasContent && !afterHasContent) { before = noteIds; after = [] }
  else if (noteIds.length > 1 && noteIdContentLengths.length === noteIds.length) {
    [before, after] = distributeNoteIdsByOverlap(noteIds, beforeContentLen, noteIdContentLengths)
  } else if (beforeContentLen >= afterContentLen) { before = noteIds; after = [] }
  else { before = []; after = noteIds }
  // Stamp SPLIT sentinels on any content-bearing side without an id so
  // save-time attribution ("where did this fresh doc come from?") is consistent.
  return [
    stampSplitSentinelIfNeeded(before, beforeHasContent),
    stampSplitSentinelIfNeeded(after, afterHasContent),
  ]
}

function stampSplitSentinelIfNeeded(ids: string[], hasContent: boolean): string[] {
  return ids.length === 0 && hasContent ? [newSentinelNoteId('split')] : ids
}

function distributeNoteIdsByOverlap(
  noteIds: string[],
  splitPos: number,
  contentLengths: number[],
): [string[], string[]] {
  const beforeIds: string[] = []
  const afterIds: string[] = []
  let offset = 0

  for (let i = 0; i < noteIds.length; i++) {
    const len = contentLengths[i]!
    const end = offset + len
    const overlapBefore = Math.max(0, Math.min(end, splitPos) - offset)
    const overlapAfter = Math.max(0, end - Math.max(offset, splitPos))

    if (overlapBefore >= overlapAfter) {
      beforeIds.push(noteIds[i]!)
    } else {
      afterIds.push(noteIds[i]!)
    }
    offset = end
  }

  return [beforeIds, afterIds]
}
